src/test/java/codeu/controller/LoginServletTest.java

```diff
@@ -53,53 +62,55 @@ public void testDoGet() throws IOException, ServletException {
   }
 
   @Test
-  public void testDoPost_BadUsername() throws IOException, ServletException {
-    Mockito.when(mockRequest.getParameter("username")).thenReturn("bad !@#$% username");
+  public void testDoPost_IncorrectPassword() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("test username");
+    Mockito.when(mockRequest.getParameter("password")).thenReturn("wrong password");
+
+    UserStore mockUserStore = Mockito.mock(UserStore.class);
+    Mockito.when(mockUserStore.isUserRegistered("test username")).thenReturn(true);
+    loginServlet.setUserStore(mockUserStore);
+    Mockito.when(mockUserStore.getUser("test username")).thenReturn(user);
+
+    HttpSession mockSession = Mockito.mock(HttpSession.class);
+    Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
 
     loginServlet.doPost(mockRequest, mockResponse);
 
-    Mockito.verify(mockRequest)
-        .setAttribute("error", "Please enter only letters, numbers, and spaces.");
-    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+    Mockito.verify(mockRequest).setAttribute("error", "Invalid password.");
   }
 
   @Test
-  public void testDoPost_NewUser() throws IOException, ServletException {
+  public void testDoPost_MissingUser() throws IOException, ServletException {
     Mockito.when(mockRequest.getParameter("username")).thenReturn("test username");
 
     UserStore mockUserStore = Mockito.mock(UserStore.class);
     Mockito.when(mockUserStore.isUserRegistered("test username")).thenReturn(false);
     loginServlet.setUserStore(mockUserStore);
+    Mockito.when(mockUserStore.getUser("test username")).thenReturn(user);
 
     HttpSession mockSession = Mockito.mock(HttpSession.class);
     Mockito.when(mockRequest.getSession()).thenReturn(mockSession);
 
     loginServlet.doPost(mockRequest, mockResponse);
 
-    ArgumentCaptor<User> userArgumentCaptor = ArgumentCaptor.forClass(User.class);
-
-    Mockito.verify(mockUserStore).addUser(userArgumentCaptor.capture());
-    Assert.assertEquals(userArgumentCaptor.getValue().getName(), "test username");
-
-    Mockito.verify(mockSession).setAttribute("user", "test username");
-    Mockito.verify(mockResponse).sendRedirect("/conversations");
+    Mockito.verify(mockRequest).setAttribute("error", "That username was not found.");
```

Comment:
What's the reason you used Mockito.verify instead of assert in this step?

Reply:
Why don't you have the old line 63 after new line 79 and 96 while having both line 115 and 116? 
``` 
Mockito.verify(mockSession).setAttribute("user", "test username");
Mockito.verify(mockResponse).sendRedirect("/conversations");

Reply:
I'm still new to this whole Mockito thing so I just copied some of the code from the previous tests. I also don't have a ton of experience with asserts, but I'm not sure how else I would access the state of mockRequest without verify.

Reply:
I think you're right to use request.verify() to check that setAttribute() is called on mockRequest as I mentioned in response to the other comment. I'm just curious if it is enough to just check setAttribute() or we have to check the same thing for getRequestDispatcher(). I tried adding 
```Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);``` 
in the last line of testDoPost_IncorrectPassword() and testDoPost_MissingUser(). The tests still pass. I think it might be safe to add the extra line for consistency and security.

Reply:
Yeah I was wondering how to check the getRequestDispatcher() for those two tests because it didn't make sense to not check if every aspect of that test was working. Thank you for finding that line!